File: extract_icons.py
#!/usr/bin/env python3
"""
Script to extract individual character icons from the AI characters sprite sheet.
Place the source image in the project root as 'ai-characters.png' or specify the path.
"""

from PIL import Image
import os
import sys

# Character names in order (left to right, top to bottom)
# Row 1: Wall-E, R2-D2, Herbie
# Row 2: C-3PO, Herbie (duplicate), EVE
# Row 3: Baymax, 7 of 9, T-800, HAL 9000, Ben 10
CHARACTER_NAMES = [
    'Wall-E', 'R2-D2', 'Herbie',
    'C-3PO', 'Herbie', 'EVE',
    'Baymax', '7 of 9', 'T-800', 'HAL 9000', 'Ben 10'
]

# Unique character names (removing duplicate Herbie)
UNIQUE_CHARACTERS = [
    'Wall-E', 'R2-D2', 'Herbie', 'C-3PO', 'EVE',
    'Baymax', '7 of 9', 'T-800', 'HAL 9000', 'Ben 10'
]

def extract_icons(image_path, output_dir='client/src/assets/ai-icons'):
    """
    Extract individual icons from the sprite sheet.
    This function will analyze the image and extract each circular icon.
    """
    # Open the image
    try:
        img = Image.open(image_path)
    except FileNotFoundError:
        print(f"Error: Image file '{image_path}' not found.")
        print("Please place the AI characters image in the project root as 'ai-characters.png'")
        sys.exit(1)
    except Exception as e:
        print(f"Error opening image: {e}")
        sys.exit(1)
    
    width, height = img.size
    print(f"Image size: {width}x{height}")
    
    # Create output directory
    os.makedirs(output_dir, exist_ok=True)
    
    # Analyze the grid layout
    # Based on the description: 3 rows, with 3, 3, and 5 icons respectively
    # We need to detect the grid automatically or use estimated positions
    
    # Estimate grid dimensions (will need to be adjusted based on actual image)
    # Assuming roughly equal spacing and circular icons
    rows = 3
    cols_row1 = 3
    cols_row2 = 3
    cols_row3 = 5
    
    # Calculate approximate icon size and spacing
    # This is a rough estimate - we'll refine based on the actual image
    estimated_icon_size = min(width // 6, height // 4)  # Rough estimate
    icon_size = estimated_icon_size
    
    # Calculate positions for each icon
    # We'll need to detect the actual positions, but for now use estimates
    icons_extracted = {}
    
    # Row 1 (3 icons)
    row1_y = height // 6  # Approximate top row position
    row1_spacing = width // 4
    
    # Row 2 (3 icons)
    row2_y = height // 2  # Middle row
    row2_spacing = width // 4
    
    # Row 3 (5 icons)
    row3_y = height * 5 // 6  # Bottom row
    row3_spacing = width // 6
    
    # Extract Row 1 icons
    for i, name in enumerate(['Wall-E', 'R2-D2', 'Herbie']):
        x = row1_spacing * (i + 1) - icon_size // 2
        y = row1_y - icon_size // 2
        if name not in icons_extracted:
            crop_box = (max(0, x), max(0, y), min(width, x + icon_size), min(height, y + icon_size))
            icon = img.crop(crop_box)
            filename = f"{name.replace(' ', '-').replace('/', '-')}.png"
            filepath = os.path.join(output_dir, filename)
            icon.save(filepath)
            icons_extracted[name] = filepath
            print(f"Extracted: {name} -> {filepath}")
    
    # Extract Row 2 icons (skip duplicate Herbie)
    row2_names = ['C-3PO', None, 'EVE']  # Skip Herbie duplicate
    for i, name in enumerate(['C-3PO', 'Herbie', 'EVE']):
        if name == 'Herbie' and 'Herbie' in icons_extracted:
            continue  # Skip duplicate
        x = row2_spacing * (i + 1) - icon_size // 2
        y = row2_y - icon_size // 2
        if name not in icons_extracted:
            crop_box = (max(0, x), max(0, y), min(width, x + icon_size), min(height, y + icon_size))
            icon = img.crop(crop_box)
            filename = f"{name.replace(' ', '-').replace('/', '-')}.png"
            filepath = os.path.join(output_dir, filename)
            icon.save(filepath)
            icons_extracted[name] = filepath
            print(f"Extracted: {name} -> {filepath}")
    
    # Extract Row 3 icons
    for i, name in enumerate(['Baymax', '7 of 9', 'T-800', 'HAL 9000', 'Ben 10']):
        x = row3_spacing * (i + 1) - icon_size // 2
        y = row3_y - icon_size // 2
        crop_box = (max(0, x), max(0, y), min(width, x + icon_size), min(height, y + icon_size))
        icon = img.crop(crop_box)
        filename = f"{name.replace(' ', '-').replace('/', '-')}.png"
        filepath = os.path.join(output_dir, filename)
        icon.save(filepath)
        icons_extracted[name] = filepath
        print(f"Extracted: {name} -> {filepath}")
    
    print(f"\nExtracted {len(icons_extracted)} unique icons to {output_dir}")
    return icons_extracted

if __name__ == '__main__':
    # Try to find the image file
    possible_paths = [
        'ai-characters.png',
        'ai-characters.jpg',
        'client/src/assets/ai-characters.png',
        'client/src/assets/ai-characters.jpg',
    ]
    
    if len(sys.argv) > 1:
        image_path = sys.argv[1]
    else:
        image_path = None
        for path in possible_paths:
            if os.path.exists(path):
                image_path = path
                break
    
    if not image_path:
        print("Error: Could not find the AI characters image.")
        print("Please either:")
        print("  1. Place the image in the project root as 'ai-characters.png'")
        print("  2. Or run: python3 extract_icons.py <path-to-image>")
        sys.exit(1)
    
    extract_icons(image_path)

File: extract_icons.js
/**
 * Node.js script to extract icons using sharp (if available) or canvas
 * Run: node extract_icons.js <image-path> [icon-size]
 */

import fs from 'fs';
import path from 'path';
import sharp from 'sharp';

// Character layout
const CHARACTER_LAYOUT = [
  ['Wall-E', 'R2-D2', 'Herbie'],
  ['C-3PO', 'Herbie', 'EVE'],
  ['Baymax', '7 of 9', 'T-800', 'HAL 9000', 'Ben 10']
];

function sanitizeFilename(name) {
  return name.replace(/\s+/g, '-').replace(/\//g, '-').replace(/ of /g, '-of-');
}

async function extractWithSharp(imagePath, outputDir) {
  try {
    const img = sharp(imagePath);
    const metadata = await img.metadata();
    
    console.log(`Image size: ${metadata.width}x${metadata.height}`);
    
    const width = metadata.width;
    const height = metadata.height;
    
    // Estimate icon size - icons appear to be roughly circular and take up about 1/5 to 1/6 of the width
    // Based on the image description, icons are circular with glows
    const estimatedIconSize = Math.floor(width / 5.5);
    const iconSize = estimatedIconSize;
    
    console.log(`Using icon size: ${iconSize}x${iconSize}`);
    
    // Create output directory
    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir, { recursive: true });
    }
    
    const iconsExtracted = {};
    
    // Row 1: 3 icons centered
    // Icons are roughly centered vertically in their row, with text below
    // Based on 3 rows, each row takes about 1/3 of height, but icons are in upper portion
    const row1Y = Math.floor(height * 0.15); // Top row, icons near top
    const row1StartX = Math.floor(width * 0.08); // Left margin
    const row1EndX = width - row1StartX; // Right margin
    const row1Spacing = Math.floor((row1EndX - row1StartX) / (CHARACTER_LAYOUT[0].length + 1));
    
    for (let i = 0; i < CHARACTER_LAYOUT[0].length; i++) {
      const name = CHARACTER_LAYOUT[0][i];
      if (!iconsExtracted[name]) {
        const centerX = row1StartX + row1Spacing * (i + 1);
        const centerY = row1Y;
        let left = Math.floor(centerX - iconSize / 2);
        let top = Math.floor(centerY - iconSize / 2);
        
        // Clamp to image bounds
        left = Math.max(0, left);
        top = Math.max(0, top);
        
        // Calculate actual extract dimensions (ensure we don't go outside bounds)
        const extractWidth = Math.min(iconSize, width - left);
        const extractHeight = Math.min(iconSize, height - top);
        
        // Validate extraction area
        if (extractWidth <= 0 || extractHeight <= 0 || left + extractWidth > width || top + extractHeight > height) {
          console.error(`⚠ Skipping ${name}: Invalid extraction area (left: ${left}, top: ${top}, width: ${extractWidth}, height: ${extractHeight}, image: ${width}x${height})`);
          continue;
        }
        
        const filename = `${sanitizeFilename(name)}.png`;
        const filepath = path.join(outputDir, filename);
        
        // Create a fresh instance for each extraction to avoid issues
        const extractImg = sharp(imagePath);
        await extractImg
          .extract({
            left: left,
            top: top,
            width: extractWidth,
            height: extractHeight
          })
          .toFile(filepath);
        
        iconsExtracted[name] = filepath;
        console.log(`✓ Extracted: ${name} -> ${filename} (${left}, ${top}, ${extractWidth}x${extractHeight})`);
      }
    }
    
    // Row 2 (skip duplicate Herbie): 3 icons
    const row2Y = Math.floor(height * 0.48); // Middle row
    const row2StartX = Math.floor(width * 0.08);
    const row2EndX = width - row2StartX;
    const row2Spacing = Math.floor((row2EndX - row2StartX) / (CHARACTER_LAYOUT[1].length + 1));
    
    for (let i = 0; i < CHARACTER_LAYOUT[1].length; i++) {
      const name = CHARACTER_LAYOUT[1][i];
      if (name === 'Herbie' && iconsExtracted['Herbie']) continue;
      if (!iconsExtracted[name]) {
        const centerX = row2StartX + row2Spacing * (i + 1);
        const centerY = row2Y;
        const left = Math.max(0, centerX - Math.floor(iconSize / 2));
        const top = Math.max(0, centerY - Math.floor(iconSize / 2));
        const extractWidth = Math.min(iconSize, width - left);
        const extractHeight = Math.min(iconSize, height - top);
        
        if (extractWidth <= 0 || extractHeight <= 0 || left >= width || top >= height) {
          console.error(`⚠ Skipping ${name}: Invalid extraction area`);
          continue;
        }
        
        const filename = `${sanitizeFilename(name)}.png`;
        const filepath = path.join(outputDir, filename);
        
        // Create a fresh instance for each extraction
        const extractImg = sharp(imagePath);
        await extractImg
          .extract({
            left: left,
            top: top,
            width: extractWidth,
            height: extractHeight
          })
          .toFile(filepath);
        
        iconsExtracted[name] = filepath;
        console.log(`✓ Extracted: ${name} -> ${filename} (${left}, ${top}, ${extractWidth}x${extractHeight})`);
      }
    }
    
    // Row 3: 5 icons (more spread out)
    const row3Y = Math.floor(height * 0.72); // Bottom row (adjusted up a bit)
    const row3StartX = Math.floor(width * 0.08); // Left margin
    const row3EndX = width - row3StartX; // Right margin
    const row3Spacing = Math.floor((row3EndX - row3StartX) / (CHARACTER_LAYOUT[2].length + 1));
    
    for (let i = 0; i < CHARACTER_LAYOUT[2].length; i++) {
      const name = CHARACTER_LAYOUT[2][i];
      const centerX = row3StartX + row3Spacing * (i + 1);
      const centerY = row3Y;
      let left = Math.floor(centerX - iconSize / 2);
      let top = Math.floor(centerY - iconSize / 2);
      
      // Clamp to image bounds
      left = Math.max(0, left);
      top = Math.max(0, top);
      
      // Calculate actual extract dimensions (ensure we don't go outside bounds)
      const extractWidth = Math.min(iconSize, width - left);
      const extractHeight = Math.min(iconSize, height - top);
      
      // Validate extraction area
      if (extractWidth <= 0 || extractHeight <= 0 || left + extractWidth > width || top + extractHeight > height) {
        console.error(`⚠ Skipping ${name}: Invalid extraction area (left: ${left}, top: ${top}, width: ${extractWidth}, height: ${extractHeight}, image: ${width}x${height})`);
        continue;
      }
      
      const filename = `${sanitizeFilename(name)}.png`;
      const filepath = path.join(outputDir, filename);
      
      // Create a fresh instance for each extraction
      const extractImg = sharp(imagePath);
      await extractImg
        .extract({
          left: left,
          top: top,
          width: extractWidth,
          height: extractHeight
        })
        .toFile(filepath);
      
      iconsExtracted[name] = filepath;
      console.log(`✓ Extracted: ${name} -> ${filename} (${left}, ${top}, ${extractWidth}x${extractHeight})`);
    }
    
    console.log(`\n✅ Successfully extracted ${Object.keys(iconsExtracted).length} unique icons to ${outputDir}/`);
    return iconsExtracted;
  } catch (error) {
    if (error.code === 'MODULE_NOT_FOUND') {
      console.error('Error: sharp module not found.');
      console.error('Install it with: npm install sharp');
      return null;
    }
    throw error;
  }
}

// Main
const imagePath = process.argv[2] || 'ai-characters.png';
const outputDir = 'client/public/assets/ai-icons';

if (!fs.existsSync(imagePath)) {
  console.error(`Error: Image file '${imagePath}' not found.`);
  console.error('Please place the AI characters image in the project root as "ai-characters.png"');
  console.error('Or run: node extract_icons.js <path-to-image>');
  process.exit(1);
}

extractWithSharp(imagePath, outputDir)
  .then(() => {
    console.log('Done!');
  })
  .catch(error => {
    console.error('Error:', error.message);
    process.exit(1);
  });

